[flang] AliasAnalysis: Handle fir.load on hlfir.designate

jdenny-ornl opened the pull request:
For example, determine that the address in `obj%p` below cannot alias the address of `v`:

```
module m
  type :: ty
    real, pointer :: p
  end type ty
end module m
subroutine test()
  use m
  real, target :: t
  real :: v
  type(ty) :: obj
  obj%p => t
  v = obj%p
end subroutine test
```

jdenny-ornl committed:
[flang] AliasAnalysis: Handle fir.load on hlfir.designate

For example, determine that the address in `obj%p` below cannot alias
the address of `v`:

```
module m
  type :: ty
    real, pointer :: p
  end type ty
end module m
subroutine test()
  use m
  real, target :: t
  real :: v
  type(ty) :: obj
  obj%p => t
  v = obj%p
end subroutine test
```

jdenny-ornl commented:
The changes in this PR are probably further motivation [for the idea of replacing the `getOriginalDef` call with a `getSource` call](https://github.com/llvm/llvm-project/pull/126156#discussion_r1953573147).  However, before that rewrite, I'm hoping this PR is simple enough that we can land it first as it is the last PR in my series and achieves an actual performance benefit.

Renaud-K commented:
It is further motivation for using getSource but I would be ok to have this change first. 

It is also further motivation to introduce a SourceKind::Emboxed because the SourceKind of the aggregate has probably nothing to do with the SourceKind of the data. 
SourceKind::Emboxed would be an indirect source kind for which we are comfortable relying on attributes to determine aliasing. 

I don't believe we can meet this guaranty for low level FIR or compiler generated FIR with non-emboxed pointers.  Meaning for instance, that some FIR could introduce aliasing without the `!fir.ptr` type. @jeanPerier , would this be accurate?

I will approve but please wait for Jean's approval as well.

Renaud-K commented:
Actually, there are a lot of TODOs.
You are getting identical source because of the reassignment of v and defOp in
```
            else if (auto allocOp =
                         def.template getDefiningOp<fir::AllocaOp>()) {
              v = def;
              defOp = v.getDefiningOp();
              type = SourceKind::Allocate;
```
which makes the aggregate the source. Would you not be getting better results commenting these 2 lines out?
```
            else if (auto allocOp =
                         def.template getDefiningOp<fir::AllocaOp>()) {
              //v = def;
              //defOp = v.getDefiningOp();
              type = SourceKind::Allocate;
```

jdenny-ornl committed:
Test multiple loads from the same alloca

jdenny-ornl commented:
> Would you not be getting better results commenting these 2 lines out?
> 
> ```
>             else if (auto allocOp =
>                          def.template getDefiningOp<fir::AllocaOp>()) {
>               //v = def;
>               //defOp = v.getDefiningOp();
>               type = SourceKind::Allocate;
> ```

I've just added new tests for allocas whose results should be MayAlias (or possibly even MustAlias with more analysis) but become NoAlias if I apply that suggested change.  For completeness, I also added similar tests for dummy args and global variables, but that suggested change only affects allocas of course.

For a dummy arg or global variable, the source of a load from it is the dummy arg or the global variable itself.  That suggested change adjusts the alloca case so the source is the load instead.  That means two different loads from the same alloca have different sources, and so we don't detect aliasing as we should.

Renaud-K commented:
Looks good. Jean is off this week but I think you can merge your change.

jdenny-ornl commented:
Thanks for the review.
